torch/csrc/utils/python_raii.h

```diff
@@ -0,0 +1,46 @@
+#include <c10/util/Optional.h>
+#include <torch/csrc/utils/pybind.h>
+#include <tuple>
+
+namespace torch {
+namespace impl {
+
+template <typename GuardT, typename... Args>
+struct RAIIContextManager {
+  explicit RAIIContextManager(Args&&... args)
+      : args_(std::forward<Args>(args)...) {}
+
+  void enter() {
+    auto emplace = [&](Args... args) {
+      return guard_.emplace(std::forward<Args>(args)...);
+    };
+    std::apply(emplace, args_);
```

Comment:
```suggestion
    std::apply(std::move(emplace), args_);
```

Reply:
Does this require the captures to be movable as well?